### PR TITLE
Virtual thread support for JVMTI extension functions

### PR DESCRIPTION
**Pull request by thallium**

### CODE REVIEW

**Comment by thallium** [comment]
@babsingh FYI

**Comment by babsingh** [comment]
@thallium This is becoming a large PR (1000+ lines). Please create a separate PR for each JVMTI extension function for the ease of reviewing and reducing risk of failures. 

When do you expect to finish this work?

Related: #18386 

fyi @pshipton @TobiAjila

**Comment by thallium** [comment]
@babsingh Currently I'm investigating the intermittent test failure of added tests so I should be able to finalize the PR next week.

Actually a lot of the changes are formatting changes and the test code for GetStackTraceExtended and GetThreadListStackTracesExtended are pretty much the same so the actual code to be reviewed is much less.

**Comment by thallium** [comment]
@babsingh The PR is ready for review. Can you please take a look?

**Comment by thallium** [comment]
5x60 tests passed https://hyc-runtimes-jenkins.swg-devops.com/view/Test_grinder/job/Grinder/38282/

File: test/functional/cmdLineTests/jvmtitests/src/com/ibm/jvmti/tests/getThreadListStackTracesExtended/gtlste002.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+		}
+
+		return rc;
+	}
+
+	public String helpMountedVthread() {
+		return "Test jvmtiGetThreadListStackTraceExtended for at least 1 jitted frame after a relatively large number of iterations on mounted virtual threads.";
+	}
+
+	public boolean testMountedVthread() {
```
Comment (babsingh) [inline test/functional/cmdLineTests/jvmtitests/src/com/ibm/jvmti/tests/getThreadListStackTracesExtended/gtlste002.java:96]: Can you also describe how the test suite invokes variants of `helpMountedVthread` and `helpMountedVthread`?

File: test/functional/cmdLineTests/jvmtitests/src/com/ibm/jvmti/tests/getStackTraceExtended/gste002.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+			if (mounted) {
+				return anyJittedFrame(null);
+			} else {
+				LockSupport.park();
+			}
+		}
+		return true;
+	}
+
+	public String helpUnmountedVthread() {
```
Comment (babsingh) [inline test/functional/cmdLineTests/jvmtitests/src/com/ibm/jvmti/tests/getStackTraceExtended/gste002.java:61]: For consistency, other methods should also have a function description similar to `methodToCompile` and `anyJittedFrame`.

**Comment by babsingh** [comment]
PR title shouldn't have a full-stop; I have removed it.

The PR description and commit message needs more details:
- which JVMTI extension functions are updated?
- what do the tests do?
- link to the related issue.

The tests should also have a comment with a high-level description.

**Comment by babsingh** [comment]
@llxia Requesting your review for the test changes.

File: test/functional/cmdLineTests/jvmtitests/playlist.xml [hunk 3]
```diff
@@ -785,4 +785,43 @@
...
+			<group>functional</group>
+		</groups>
+		<types>
+			<type>native</type>
+		</types>
+		<versions>
+			<version>21+</version>
+		</versions>
+		<features>
+			<feature>AOT:explicit</feature>
```
Comment (llxia) [inline test/functional/cmdLineTests/jvmtitests/playlist.xml:820]: Is this AOT feature? See doc for AOT feature tag: https://github.com/eclipse-openj9/openj9/blob/master/test/docs/OpenJ9TestUserGuide.md#run-tests-against-specific-feature

File: test/functional/cmdLineTests/jvmtitests/src/com/ibm/jvmti/tests/getThreadListStackTracesExtended/gtlste002.java (hunk 1 above)
Comment (thallium) [reply test/functional/cmdLineTests/jvmtitests/src/com/ibm/jvmti/tests/getThreadListStackTracesExtended/gtlste002.java:96]: Any function whose name starts with "test" will be called by the test suite.

File: test/functional/cmdLineTests/jvmtitests/playlist.xml (hunk 3 above)
Comment (thallium) [reply test/functional/cmdLineTests/jvmtitests/playlist.xml:820]: Thanks for pointing this out. Also, is there documentation on the meaning of different modes? Might need to update modes as well.

**Comment by babsingh** [comment]
There are typos in the commit messages:

```
In each of the tests, a virtual thread runs methodToCompile()
for a relatively large number of iterations to JIT compile the
method. Then, we use the JVMTI Extension API to check if there
are any JITted frames in the stack trace of the virtual thread.
We test this for both mounted and unmounted virtual threads.
```

**Comment by babsingh** [comment]
@llxia There are two variants of existing tests in `jvmtitests/playlist.xml`: `^arch.390` and `arch.390`. Do we need to do the same for the new test?

File: test/functional/cmdLineTests/jvmtitests/playlist.xml (hunk 3 above)
Comment (llxia) [reply test/functional/cmdLineTests/jvmtitests/playlist.xml]: For modes, you can see https://github.com/adoptium/TKG/blob/master/resources/modes.xml

**Comment by babsingh** [comment]
> The new test cmdLineTester_jvmtitests_Java21andUp only uses jvmtitests_Java21andUp.xml that you added. Do we want it to run on arch.390?

It should be run everywhere. But, existing tests have two targets: one for arch.390 (e.g. cmdLineTester_jvmtitests_nongold) and another for non arch.390 platforms (e.g cmdLineTester_jvmtitests). The only visible difference is that the arch.390 test target is tagged to extended.functional whereas the non arch.390 test target is tagged to sanity.functional. Not sure why this was done. If it doesn't matter anymore, we can just have one test target.

**Comment by thallium** [comment]
I've updated the test variations to test different GC policies. Do we need to test compressed ref as well?

**Comment by babsingh** [comment]
>  Do we need to test compressed ref as well?

The current settings have compressed refs enabled. 

These changes LGTM. I plan to launch the PR builds. @llxia Do the test changes look good to you?

**Comment by babsingh** [comment]
jenkins test sanity.functional,extended.functional zlinux jdk11

**Comment by babsingh** [comment]
jenkins test sanity.functional,extended.functional,sanity.openjdk,extended.openjdk win,alinux jdk21

**Comment by babsingh** [comment]
jenkins test extended.functional,sanity.openjdk,extended.openjdk amac jdk22

**Comment by babsingh** [comment]
See existing tests such as `soae001` (Java11andUp). One missing item is
```
#if JAVA_SPEC_VERSION >= 21
   { "gste002", gste002, "com.ibm.jvmti.tests.getStackTraceExtended.gste002", "GetStackTraceExtended" },
   { "gtlste002", gtlste002, "com.ibm.jvmti.tests.getThreadListStackTracesExtended.gtlste002", "GetThreadListStackTracesExtended" },
#endif /* JAVA_SPEC_VERSION >= 21 */
```

File: runtime/tests/jvmtitests/module.xml [hunk 4]
```diff
@@ -47,8 +47,10 @@
 		<export name="Java_com_ibm_jvmti_tests_getStackTrace_gst001_check"/>
 		<export name="Java_com_ibm_jvmti_tests_getStackTrace_gst002_check"/>
 		<export name="Java_com_ibm_jvmti_tests_getStackTraceExtended_gste001_anyJittedFrame"/>
+		<export name="Java_com_ibm_jvmti_tests_getStackTraceExtended_gste002_anyJittedFrame"/>
 		<export name="Java_com_ibm_jvmti_tests_getAllStackTracesExtended_gaste001_anyJittedFrame"/>
 		<export name="Java_com_ibm_jvmti_tests_getThreadListStackTracesExtended_gtlste001_anyJittedFrame"/>
+		<export name="Java_com_ibm_jvmti_tests_getThreadListStackTracesExtended_gtlste002_anyJittedFrame"/>
```
Comment (babsingh) [inline runtime/tests/jvmtitests/module.xml:53]: this needs `<exports group="jdk21">`. see the `soae001` code below.

File: runtime/tests/jvmtitests/src/com/ibm/jvmti/tests/getStackTraceExtended/gste002.c [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+#include <stdlib.h>
+#include <string.h>
+
+#include "ibmjvmti.h"
+#include "jvmti_test.h"
+
+#define JVMTI_TEST_GSTE002_MAX_FRAMES 50
+
+static agentEnv *env = NULL;
+
```
Comment (babsingh) [inline runtime/tests/jvmtitests/src/com/ibm/jvmti/tests/getStackTraceExtended/gste002.c:31]: the entire files needs to be wrapped with `#if JAVA_SPEC_VERSION >= 21`. see `soae001.c`.

File: runtime/tests/jvmtitests/src/com/ibm/jvmti/tests/getThreadListStackTracesExtended/gtlste002.c [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+#include <stdlib.h>
+#include <string.h>
+
+#include "ibmjvmti.h"
+#include "jvmti_test.h"
+
+#define JVMTI_TEST_GTLSTE002_MAX_FRAMES 50
+
+static agentEnv *env = NULL;
+
```
Comment (babsingh) [inline runtime/tests/jvmtitests/src/com/ibm/jvmti/tests/getThreadListStackTracesExtended/gtlste002.c:31]: the entire files needs to be wrapped with `#if JAVA_SPEC_VERSION >= 21`. see `soae001.c`.

**Comment by thallium** [comment]
@babsingh all the above changes are mostly related to the native functions. We haven't add anything to prevent the Java testing code from compiling. In the `playlist.xml` we already have:
```xml
		<versions>
			<version>21+</version>
		</versions>
```
but that doesn't seem to be effective. I'm stilling getting the same error in my local testing.

**Comment by babsingh** [comment]
In `.../jvmtitests/build.xml`, adding `<exclude name="...` for those files might prevent them from compiling. @llxia for confirming?

**Comment by llxia** [comment]
playlist.xml controls the execution only. For java compilation, we need to update build.xml.


```
<if>
<not>
   <equals arg1="${JDK_VERSION}" arg2="11" />
</not>
    //add additional exclude files
       ...
```
https://github.com/eclipse-openj9/openj9/blob/6b03a382a12a2293d1cb4d884c55899859da516e/test/functional/cmdLineTests/jvmtitests/build.xml#L81

**Comment by thallium** [comment]
I've added exclude files to build.xml.

File: test/functional/cmdLineTests/jvmtitests/build.xml [hunk 7]
```diff
@@ -75,10 +75,24 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-ex
 						<property name="excludeFile" value="" />
 					</else>
 				</if>
+				<!-- Java21 and up tests-->
+				<if>
+					<matches string="${JDK_VERSION}" pattern="^(2[1-9]|[3-9][0-9])$$" />
```
Comment (llxia) [inline test/functional/cmdLineTests/jvmtitests/build.xml:80]: We can do
```
			<not>
				<matches string="${JDK_VERSION}" pattern="^(8|11|17)$$" />
			</not>
```

**Comment by babsingh** [comment]
jenkins test extended.functional zlinux jdk11

**Comment by babsingh** [comment]
jenkins test extended.functional,extended.openjdk win,alinux jdk21

**Comment by babsingh** [comment]
jenkins test extended.functional,extended.openjdk amac jdk22

**Comment by babsingh** [comment]
LGTM. @llxia I will merge the PR after you complete your review.

> https://openj9-jenkins.osuosl.org/job/Test_openjdk21_j9_extended.openjdk_x86-64_windows_Personal/3

An infra issue similar to https://github.com/eclipse-openj9/openj9/issues/8446 is seen in the above PR build. fyi @AdamBrousseau

**Review by babsingh — APPROVED** [review]

**Review by llxia — APPROVED** [review]
The test change looks good to me

File: test/functional/cmdLineTests/jvmtitests/build.xml [hunk 8]
```diff
@@ -75,10 +75,26 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-ex
 						<property name="excludeFile" value="" />
 					</else>
 				</if>
+				<!-- Java21 and up tests-->
+				<if>
+					<not>
+						<matches string="${JDK_VERSION}" pattern="^(8|11|17)$$" />
+					</not>
```
Comment (keithc-ca) [inline test/functional/cmdLineTests/jvmtitests/build.xml:82]: This doesn't work for Java 8 because that version is handled above by lines 57-63.

Comment (babsingh) [reply test/functional/cmdLineTests/jvmtitests/build.xml:82]: Fixed in https://github.com/eclipse-openj9/openj9/pull/19183.

**Pull request merged**